# Machine Learning Intern at InfoPillar Solutions Pvt Ltd - Virtual Internship Program

# Author - Shubham Shriram Chaudhari

# Task 1 :- Fake News Detection Project

# 1] Import Libraries

In [2]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# 2] Load Dataset

In [3]:
#Read the data
dataset = pd.read_csv('news.csv')

# 3] Summarise Dataset

In [4]:
print("No. of Columns and Rows :- ",dataset.shape) #It gives no. of rows and columns

print(dataset.head()) #It gives first five record

No. of Columns and Rows :-  (6335, 4)
   Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL  


# 4] Get the Labels from DataSet

In [5]:
labels=dataset.label
# labels.head(10)
print(labels)

0       FAKE
1       FAKE
2       REAL
3       FAKE
4       REAL
        ... 
6330    REAL
6331    FAKE
6332    FAKE
6333    REAL
6334    REAL
Name: label, Length: 6335, dtype: object


# 5] Split the dataset into training and testing sets

In [6]:
#Spliting the dataset into testing and training
x_train,x_test,y_train,y_test =train_test_split(dataset['text'], labels, test_size=0.2, random_state=7)

# 6] Fit and transform the vectorizer on the train set, and transform the vectorizer on the test set.

In [7]:
#Initialize a TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

In [8]:
#Fit and transform train set, transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

# 7] Initialize a PassiveAggressiveClassifier.

In [9]:
#Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

PassiveAggressiveClassifier(max_iter=50)

In [10]:
#Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 93.13%


# The accuracy of 93.13% with this model.

# 8]Print out a confusion matrix.

In [11]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[591,  47],
       [ 40, 589]], dtype=int64)

# In this model, we have 591 true positives, 589 true negatives, 47 false positives, and 40 false negatives.
